### PR TITLE
Update sparse.grad() to support re-packing gradients into PyTrees

### LINKED ISSUE #16582: `sparse.grad` only returns the gradient with respect to the first element of a PyTree
### Description

When applying `sparse.grad` from `jax.experimental.sparse` to a function which take in a Pytree as the first argument, only the gradient with respect to the first item in the Pytree is returned. This is both unexpected and inconsistent with the behaviour of `jax.grad`, which returns a gradient which has the same tree structure as the input. 

Here is a small working example demonstrating this behaviour:

```python
import jax
import jax.numpy as jnp
from jax.experimental import sparse

def foo1(wb, x, y): # first item is a tuple with W first and B second
    w, b = wb
    return ((w @ x + b) - y).sum()

def foo2(bw, x, y): # first item is a tuple with B first and W second
    b, w = bw
    return ((w @ x + b) - y).sum()

rng = jax.random.PRNGKey(0)
keys = jax.random.split(rng, 4)
w = jax.random.normal(keys[0], (3, 3))
b = jax.random.normal(keys[1], (3,))
x = jax.random.normal(keys[2], (3,))
y = jax.random.normal(keys[3], (3,))
"""
```

Here are the outputs for the different tests:

```python
# normal jax.grad, W first and B second

jax.grad(foo1)((w, b), x, y)

(Array([[-0.47994015,  0.42577833,  0.765658  ],
        [-0.47994015,  0.42577833,  0.765658  ],
        [-0.47994015,  0.42577833,  0.765658  ]], dtype=float32),
 Array([1., 1., 1.], dtype=float32))

# ============================

# normal jax.grad, B first and W second

jax.grad(foo2)((b, w), x, y)

(Array([1., 1., 1.], dtype=float32),
 Array([[-0.47994015,  0.42577833,  0.765658  ],
        [-0.47994015,  0.42577833,  0.765658  ],
        [-0.47994015,  0.42577833,  0.765658  ]], dtype=float32))

# ============================

# sparse.grad, W first and B second. only the gradient with respect to W is returned!

sparse.grad(foo1)((w, b), x, y)

Array([[-0.47994015,  0.42577833,  0.765658  ],
       [-0.47994015,  0.42577833,  0.765658  ],
       [-0.47994015,  0.42577833,  0.765658  ]], dtype=float32)

# ============================

# sparse.grad, B first and W second. only the gradient with respect to B is returned!

sparse.grad(foo2)((b, w), x, y)

Array([1., 1., 1.], dtype=float32)
```

### What jax/jaxlib version are you using?

jax v0.4.13

### Which accelerator(s) are you using?

_No response_

### Additional system info

_No response_

### NVIDIA GPU info

_No response_

### PR DESCRIPTION
**Pull request by Blair-Johnson**
Fixes #16582

This PR modifies the postprocessing step of `sparse.grad()` to reconstruct the input PyTrees that were indexed for autodiff. Previously, only the gradient corresponding to the first element of a PyTree would be returned. 

This PR includes several test cases to verify the behavior of `sparse.grad()` matches that of `jax.grad()` when gradients are taken with respect to pytrees.

### CODE REVIEW

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: tests/sparse_test.py [hunk 1]
```diff
@@ -725,6 +725,102 @@ def f(X, y):
...
+        grad_de = jax.grad(g, argnums=1, has_aux=has_aux)(y, Xtree_de)
+        grad_sp = sparse.grad(g, argnums=1, has_aux=has_aux)(y, Xtree_sp)
+
+      if has_aux:
+        grad_de, aux_de = grad_de
+        grad_sp, aux_sp = grad_sp
+        self.assertAllClose(aux_de, aux_sp)
+
+      # Pytree structure
+      def map_todense(x):
```
[the remark this answers is not in the document]
Comment (Blair-Johnson) [reply tests/sparse_test.py:789]: Nice, I didn't know that.

**Review by jakevdp — APPROVED** [review]

File: jax/experimental/sparse/ad.py [hunk 2]
```diff
@@ -66,9 +66,16 @@ def reconstruct(i, grad_out):
       f_recons = jax.vmap(f_recons)
     return f_recons(grad_out)
 
+  def tree_is_none(tree):
+    return tree_util.tree_all(tree_util.tree_map(lambda x: x is None, tree))
```
Comment (jakevdp) [inline jax/experimental/sparse/ad.py:70]: Since `None` represents an empty subtree and not a leaf, a more efficient way to do this is `tree_is_none = lambda tree: not jax.tree.leaves(tree)`

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/experimental/sparse/ad.py [hunk 3]
```diff
@@ -66,9 +66,16 @@ def reconstruct(i, grad_out):
...
+
   def postprocess_gradients(grads_out):
-    out = [reconstruct(*args) for args in safe_zip(argnums_flat1, grads_out)]
-    return out[0] if isinstance(argnums, int) else out
+    grads = [reconstruct(*args) for args in safe_zip(argnums_flat1, grads_out)]
+    leaf_grads = [grads.pop(0) if l_idx in argnums_flat1 else None for l_idx in
```
Comment (jakevdp) [inline jax/experimental/sparse/ad.py:74]: This is confusing (and potentially error-prone, if `grads` is not the expected size). What about something like this:
```python
leaf_grads = [None] * tree1.num_leaves
for i, grad in safe_zip(argnums_flat_1, grads):
  leaf_grads[i] = grad
```

File: jax/experimental/sparse/ad.py [hunk 4]
```diff
@@ -66,9 +66,16 @@ def reconstruct(i, grad_out):
...
+                  range(tree1.num_leaves)]
+    grad_tree = tree_util.tree_unflatten(tree1, leaf_grads)
+    grad_tree = tuple(filter(lambda x: not tree_is_none(x), grad_tree))
```
Comment (jakevdp) [inline jax/experimental/sparse/ad.py:77]:
```suggestion
    grad_tree = tuple(filter(lambda x: jax.tree.leaves(x), grad_tree))
```

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/experimental/sparse/ad.py (hunk 3 above)
Comment (Blair-Johnson) [reply jax/experimental/sparse/ad.py]: This is a good improvement; I've incorporated it.

**Comment by Blair-Johnson** [comment]
@jakevdp Finally getting back around to this, apologies for the glacial pace. I adopted your suggestions with some minor fixes.

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by jakevdp** [comment]
Hi - we're seeing a couple test failures here that have been fixed on the main branch. Can you rebase against the most recent main branch commit so we can try again?
